File: src/main/java/de/maggiwuerze/xdccloader/events/CustomSpringEventListener.java
package de.maggiwuerze.xdccloader.events;

import de.maggiwuerze.xdccloader.events.download.DownloadCreationEvent;
import de.maggiwuerze.xdccloader.events.download.DownloadDeleteEvent;
import de.maggiwuerze.xdccloader.events.download.DownloadDoneEvent;
import de.maggiwuerze.xdccloader.events.download.DownloadUpdateEvent;
import de.maggiwuerze.xdccloader.irc.IrcBot;
import de.maggiwuerze.xdccloader.irc.IrcEventListener;
import de.maggiwuerze.xdccloader.model.download.Download;
import de.maggiwuerze.xdccloader.model.download.DownloadState;
import de.maggiwuerze.xdccloader.model.entity.Bot;
import de.maggiwuerze.xdccloader.service.DownloadService;
import de.maggiwuerze.xdccloader.service.IrcBotService;
import de.maggiwuerze.xdccloader.util.IpHelper;
import de.maggiwuerze.xdccloader.util.ProgressWatcherFactory;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.RandomStringUtils;
import org.pircbotx.Configuration;
import org.pircbotx.exception.IrcException;
import org.springframework.context.event.EventListener;
import org.springframework.core.task.SimpleAsyncTaskExecutor;
import org.springframework.core.task.TaskExecutor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
@Slf4j
public class CustomSpringEventListener {

	private final ProgressWatcherFactory progressWatcherFactory;
	private final IrcEventListener ircEventListener;
	private final EventPublisher eventPublisher;
	private final DownloadService downloadService;
	private final IrcBotService ircBotService;

	@EventListener
	public void onDownloadCreationEvent(DownloadCreationEvent event) {

		Download download = downloadService.getById(event.getPayload());
		eventPublisher.updateDownloadState(DownloadState.PREPARING, download);
		//Create our bot with the configuration
		download.setProgressWatcher(progressWatcherFactory.getProgressWatcher(download.getId()));

		IrcBot bot = ircBotService.getIrcBotForDownload(download);

		eventPublisher.updateDownloadState(DownloadState.PREPARED, download);
		TaskExecutor taskExecutor = new SimpleAsyncTaskExecutor(bot.getNick() + "_bot");
		taskExecutor.execute(() -> {

			try {

				bot.startBot();
			} catch (IOException | IrcException e) {

				log.error(e.toString());
			}
		});

		eventPublisher.updateDownloadState(DownloadState.CONNECTING, download);
	}

	@EventListener
	public void onDownloadUpdateEvent(DownloadUpdateEvent event) {

		downloadService.update(downloadService.getById(event.getPayload()));
		eventPublisher.sendWebsocketEvent(SocketEvents.UPDATED_DOWNLOAD, downloadService.getById(event.getPayload()));
	}

	@EventListener
	public void onDownloadDeleteEvent(DownloadDeleteEvent event) {

		eventPublisher.sendWebsocketEvent(SocketEvents.DELETED_DOWNLOAD, downloadService.getById(event.getPayload()));
		downloadService.remove(event.getPayload());
	}

	@EventListener
	public void onDownloadDoneEvent(DownloadDoneEvent event) {

		//Do some stuff here before setting to done!
		Download download = downloadService.getById(event.getPayload());
		download.setStatus(DownloadState.DONE);
		downloadService.update(download);

		eventPublisher.sendWebsocketEvent(SocketEvents.UPDATED_DOWNLOAD, downloadService.getById(event.getPayload()));
	}
}

File: src/main/kotlin/de/maggiwuerze/xdccwebloader/events/CustomSpringEventListener.kt
package de.maggiwuerze.xdccwebloader.events

import de.maggiwuerze.xdccwebloader.events.download.DownloadCreationEvent
import de.maggiwuerze.xdccwebloader.events.download.DownloadDeleteEvent
import de.maggiwuerze.xdccwebloader.events.download.DownloadDoneEvent
import de.maggiwuerze.xdccwebloader.events.download.DownloadUpdateEvent
import de.maggiwuerze.xdccwebloader.irc.IrcBot
import de.maggiwuerze.xdccwebloader.model.download.DownloadState
import de.maggiwuerze.xdccwebloader.service.DownloadService
import de.maggiwuerze.xdccwebloader.service.IrcBotService
import de.maggiwuerze.xdccwebloader.util.ProgressWatcherFactory
import org.pircbotx.exception.IrcException
import org.slf4j.LoggerFactory
import org.springframework.context.event.EventListener
import org.springframework.core.task.SimpleAsyncTaskExecutor
import org.springframework.core.task.TaskExecutor
import org.springframework.stereotype.Component
import java.io.IOException

@Component
class CustomSpringEventListener(
    val progressWatcherFactory: ProgressWatcherFactory,
    val eventPublisher: EventPublisher,
    val downloadService: DownloadService,
    val ircBotService: IrcBotService
) {

    val log = LoggerFactory.getLogger(this.javaClass.name)

    @EventListener
    fun onDownloadCreationEvent(event: DownloadCreationEvent) {
        downloadService.getOrThrow(event.payload).let { download ->
            eventPublisher.updateDownloadState(DownloadState.PREPARING, download)
            //Create our bot with the configuration
            download.progressWatcher = progressWatcherFactory.getProgressWatcher(download.id)

            val bot: IrcBot = ircBotService.getIrcBotForDownload(download)

            eventPublisher.updateDownloadState(DownloadState.PREPARED, download)
            val taskExecutor: TaskExecutor = SimpleAsyncTaskExecutor(bot.getNick() + "_bot")
            taskExecutor.execute {
                try {
                    bot.startBot()
                } catch (e: IOException) {
                    log.error(e.toString())
                } catch (e: IrcException) {
                    log.error(e.toString())
                }
            }

            eventPublisher.updateDownloadState(DownloadState.CONNECTING, download)
        }
    }

    @EventListener
    fun onDownloadUpdateEvent(event: DownloadUpdateEvent) {
        downloadService.getOrThrow(event.payload).let { download ->
            eventPublisher.sendWebsocketEvent(SocketEvents.UPDATED_DOWNLOAD, download)
        }
    }

    @EventListener
    fun onDownloadDeleteEvent(event: DownloadDeleteEvent) {
        downloadService.getOrThrow(event.payload).let { download ->
            eventPublisher.sendWebsocketEvent(SocketEvents.DELETED_DOWNLOAD, download)
            downloadService.remove(event.payload)
        }
    }

    @EventListener
    fun onDownloadDoneEvent(event: DownloadDoneEvent) {
        //Do some stuff here before setting to done!

        downloadService.getOrThrow(event.payload).let { download ->
            download.status = DownloadState.DONE
            downloadService.update(download)

            eventPublisher.sendWebsocketEvent(SocketEvents.UPDATED_DOWNLOAD, download)
        }
    }
}